lib/ace/ext/papeeria/tex_completer.coffee

```diff
@@ -86,37 +93,68 @@ define( (require, exports, module) ->
     meta: 'equation'
   )
   
-  init = (editor, bindKey) ->
+  #Specific for token's system of type in ace
+  #We saw such a realization in html_completions.js
+  istype = (token, type) -> 
+    return token.type.lastIndexOf(type) > -1
+  
+
+  init = (editor, bindKey) -> 
     HashHandler = require("ace/keyboard/hash_handler").HashHandler; 
     keyboardHandler = new HashHandler();
     keyboardHandler.addCommand(
       name: 'add item in list mode',
       bindKey: bindKey,
-      exec: (editor) ->
+      exec: (editor) ->  
         pos = editor.getCursorPosition()
         curLine = editor.session.getLine(pos.row);
+        intendCount = ContextHelper.getNestedListDepth(editor.session, pos.row)
         # it's temporary fix bug with added \item before \begin{itemize|enumerate}
-        if (editor.getSession().getContext(pos.row) == "list" && curLine.indexOf("begin") < pos.column) 
-          editor.insert("\n\t\\item ")
+        if ContextHelper.getContext(editor.session, pos.row) == LIST_STATE && curLine.indexOf("begin") < pos.column
+          editor.insert("\n" + "    ".repeat(intendCount) + "\\item ")
           return true
         else 
           return false
-    )
-    editor.keyBinding.addKeyboardHandler(keyboardHandler)
+        
+    );
+    editor.keyBinding.addKeyboardHandler(keyboardHandler);
 
-  getCompletions = (editor, session, pos, prefix, callback) ->
-    context = session.getContext(pos.row)
-    if context == "start"
-      callback(null, listSnippets.concat(equationSnippets.concat(basicSnippets)))
+  class ReferenceGetter
+    constructor: ->
+      @cachedURL =  ""
+      @cache = []
+    processData: (data) -> 
+      @cache = data.map((elem) => 
+                  return {
+                    name: elem.caption
+                    value: elem.caption
+                    score: Number.MAX_VALUE
+                    meta: "ref"
+                  }
+            )
 
-    if context == LIST_STATE
-      callback(null, listKeywords_.concat(listSnippets.concat(equationSnippets)))
+    getReference: (url, callback) -> 
+      if (url != @cashedURL)
+        @cachedURL = url
+        $.getJSON(url).done((data) => @processData(data); callback(null, @cache))
 
-    if context == EQUATION_STATE
-      callback(null, formulasSnippets.concat(equationKeywords_))
-  
+  class TexCompleter
+      constructor: ->
+        @refGetter = new ReferenceGetter()
+      getCompletions: (editor, session, pos, prefix, callback) ->
```

Comment:
Use thick arrow, otherwise chances are that `@refGetter` below will be undefined